File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/TopUpKagemushaV1Instruction.java
// Copyright 2026 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.android.model.instructions;

import java.util.Objects;
import org.hyperledger.iroha.android.model.InstructionBox;
import org.hyperledger.iroha.android.offline.KagemushaNoritoV1;
import org.hyperledger.iroha.sdk.offline.KagemushaTopUpRequestV1;

/** Native first-release instruction for one payer-authorized KAGEMUSHA reserve top-up. */
public final class TopUpKagemushaV1Instruction {
  /** Sole first-release dynamic instruction registry identifier. */
  public static final String WIRE_ID = "iroha.kagemusha.v1.top_up";

  /** Exact concrete Rust type whose schema hash binds the instruction payload. */
  public static final String SCHEMA_NAME =
      "iroha_data_model::isi::kagemusha_v1::TopUpKagemushaV1";

  private final KagemushaTopUpRequestV1 request;

  public TopUpKagemushaV1Instruction(final KagemushaTopUpRequestV1 request) {
    this.request = Objects.requireNonNull(request, "request");
    KagemushaNoritoV1.encodeTopUpRequestShape(request);
  }

  /** Returns the complete proof-bearing deterministic top-up intent. */
  public KagemushaTopUpRequestV1 request() {
    return request;
  }

  /** Returns the registered, schema-bound instruction ready for a transaction executable. */
  public InstructionBox toInstructionBox() {
    return InstructionBox.fromWirePayload(
        WIRE_ID, KagemushaNoritoV1.encodeTopUpInstructionPayloadShape(request));
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/TopUpKagemushaV1Instruction.kt
// Copyright 2026 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.sdk.core.model.instructions

import org.hyperledger.iroha.sdk.core.model.InstructionBox
import org.hyperledger.iroha.sdk.offline.KagemushaNoritoV1
import org.hyperledger.iroha.sdk.offline.KagemushaTopUpRequestV1

/**
 * Native first-release instruction for one payer-authorized KAGEMUSHA reserve top-up.
 *
 * Shape validation does not grant monetary authority; Core verifies the embedded authorization
 * before debiting the transaction authority.
 */
class TopUpKagemushaV1Instruction(
    @JvmField val request: KagemushaTopUpRequestV1,
) {
    init {
        KagemushaNoritoV1.encodeTopUpRequestShape(request)
    }

    /** Return the registered, schema-bound instruction ready for a transaction executable. */
    fun toInstructionBox(): InstructionBox =
        InstructionBox.fromWirePayload(
            WIRE_ID,
            KagemushaNoritoV1.encodeTopUpInstructionPayloadShape(request),
        )

    companion object {
        /** Sole first-release dynamic instruction registry identifier. */
        const val WIRE_ID: String = "iroha.kagemusha.v1.top_up"

        /** Exact concrete Rust type whose schema hash binds the instruction payload. */
        const val SCHEMA_NAME: String =
            "iroha_data_model::isi::kagemusha_v1::TopUpKagemushaV1"
    }
}
